### PR TITLE
tests: update permission denied message for test-snapd-event on ubuntu 2104

### PR DESCRIPTION
**Pull request by sergiocazzolato**
Update the message checked when the permission is denied in test
tests/main/security-dev-input-event-denied

The denial is the folowing:
[  608.831636] audit: type=1400 audit(1614200822.854:59):
apparmor="DENIED" operation="open"
profile="snap.test-snapd-event.test-snapd-event"
name="/dev/input/event2" pid=32386 comm="read-evdev-devi"
requested_mask="wrc" denied_mask="wrc" fsuid=0 ouid=0

The error on the test is:
+ test-snapd-event -event-kbd
+ MATCH 'Permission denied'
grep error: pattern not found, got:
/snap/test-snapd-event/x1/bin/read-evdev-device: line 27:
/dev/input/event2: Operation not permitted

### CODE REVIEW

File: tests/main/security-dev-input-event-denied/task.yaml [hunk 1]
```diff
@@ -50,7 +50,7 @@ execute: |
         exit 1
     fi
     # AppArmor is 'Permission denied' which is expected with default policy
-    MATCH "Permission denied" < call.error
+    MATCH "(Permission denied|Operation not permitted)" < call.error
```
Comment (bboozzoo) [inline tests/main/security-dev-input-event-denied/task.yaml:53]: This is unexpected. When AppArmor denies an operation it returns EACCESS (Permission denied), but in this case we're getting EPERM, suggesting that the call was blocked for some other reason. I don't think we should do this change without understanding why the error is different.

Comment (sergiocazzolato) [reply tests/main/security-dev-input-event-denied/task.yaml:53]: @bboozzoo, thanks for the review, agree with you, I'll run this in a 20.04 and compare the dmesg info, perhaps that could help.

Comment (sergiocazzolato) [reply tests/main/security-dev-input-event-denied/task.yaml:53]: @bboozzoo I found this: https://paste.ubuntu.com/p/jVYtFCvbgP/

Comment (anonymouse64) [reply tests/main/security-dev-input-event-denied/task.yaml:53]: @sergiocazzolato actually the messages / returned error messages from that pastebin make sense and are correct, when the joystick interface is not connected, the first error message should be "Permission Denied", and the second error message when the joystick interface _is connected_ should be "Operation not permitted"

Comment (anonymouse64) [reply tests/main/security-dev-input-event-denied/task.yaml:53]: when this test fails, what I think it indicates is that either:
* the apparmor message for the failure is wrong and for some reason appamor is causing EPERM due to some bug there
* the generated policy is wrong and we are somehow allowing access to whatever device node is referenced by the `-event-kbd` argument to evdev
* the device node referenced by the `-event-kbd` argument to evdev actually shows up somewhere else, and is not what we think it is anymore (i.e. maybe it's not /dev/input/event* but it's /dev/event* or some other odd thing

**Review by anonymouse64 — CHANGES_REQUESTED** [review]
I think the way forward here given the fact that:

1) snapd does not remove device cgroups for snaps even when they should not be put into device cgroups because the interfaces necessitating the device cgroup are not connected
2) hirsute specifically suffers from the bug where when a snap has both a device cgroup and an apparmor profile active the device cgroup always gets evaluated first 

we should do the following:

1) on all systems, remove the snap from the device cgroup when the relevant interfaces are disconnected, this can be done with i.e. `rmdir /sys/fs/cgroup/devices/snap.test-snapd-event`. Also leave a comment about https://bugs.launchpad.net/snapd/+bug/1803210 when this is done.
2) on restore of the test (or arguably generically on restore of _every_ test), unconditionally remove all related cgroups for all snaps - this does not effect any non-running snap, since if we remove the device cgroup when the snap should have one, snap-confine will just re-create the device cgroup when it runs again, and there should be no running snaps after each test restore
3) we should also add an invariant check to the start/beginning of every test via tests.invariant (much like the existing ones for crashed snap-confine) that ensures that there are no leftover running snap processes, now that we have cgroup tracking this can be checked for every user on ubuntu at least with `cat /sys/fs/cgroup/unified/user.slice/user-*.slice/user@*.service/snap.*/cgroup.procs` (this path will be different on cgroup v2 only systems and cgroup v1 only systems, IIRC ubuntu groovy is hybrid v2). This will ensure that there are no leftover processes from snaps running, and also ensures that we can safely implement 2 without raising privileges for any running snap processes the test might be executing.

Implementing all of the above means we don't care about the order in which device cgroup / apparmor is evaluated in the kernel, since we will only ever have 1 effective enforcement mechanism that denies the access in place at any given point of the test where we care about the error message:

* when the interface is disconnected, we don't have the device cgroup in place and we will not have apparmor policy allowing access to the path /dev/input/event*, hence the only denial we should see is from apparmor
* when the interface is connected, we will have policy in apparmor which will allow the path /dev/input/... access through always, but we will have the device cgroup enforcement in effect denying access to non-device-buttons and non-joystick device nodes in /dev/input/event*, so the message we see should always be from the device cgroup

This also means we can avoid hard-coding platform specific behavior like we see on Hirsute right now, which is a plus in my mind.

**Comment by sergiocazzolato** [comment]
@anonymouse64 thanks for the explanation, I'll work on that fix. 
Thanks

**Comment by sergiocazzolato** [comment]
@anonymouse64, I manually deleted the path and also by using the cgdelete tool: https://paste.ubuntu.com/p/9Js99qX9xQ/ 
In both cases, I see the same behavior

**Comment by anonymouse64** [comment]
@sergiocazzolato I'm working on a workaround for the problem you describe, indeed it is more complicated and we need to untag the devices not just remove the device cgroup...

**Review by anonymouse64 — APPROVED** [review]
Unfortunately, I can't seem to get my solution to work here. My thinking was the following:

1) leave the joystick interface connected 
1) modify the rules file that gets executed to instead of adding a tag, to remove the tag with `TAG-=...`
1) do `udevadm settle` to trigger the rule to be run again, this time the devices will have the tag removed for the snap under test in the udev db
1) disconnect the interface
1) remove the cgroup manually
1) re-execute the snap, this time, there should not be any cgroup created since we untagge the device

Unfortunately I can't ever seem to get step 3 to work, as it seems the devices continued to be tagged as associated with the snap's security tag, and thus when we re-execute the snap, snap-confine consults the udev db and finds that there are devices tagged for the security tag of the snap, and thus re-creates the cgroup and adds the devices to that cgroup, which then leads us to the fact that on Hirsute, the device cgroup LSM is always processed first, before the apparmor security LSM, leading to always getting the "Operation not Permitted" message. 

As such, since we need to make hirsute required since it is now released and we don't want to regress there at all, I'm approving this

**Comment by bboozzoo** [comment]
I think I linked those in the standup:
- https://github.com/systemd/systemd/issues/4214
- https://github.com/systemd/systemd/pull/16853

The notes mention udev_device_has_current_tag(). AFAIU the current tags can be listed starting from systemd 247 and guess which version is in 21.04 :smile: Since this sounds like a libudev API change too, and we build the snapd snap on 16.04, we'll need to get a bit creative to workaround this. Since tags are sticky, we'll likely need to leave the enumeration code as it is now, but apply more filtering by inspecting the CURRENT_TAGS property ourselves?

**Comment by bboozzoo** [comment]
Looking a bit more, on hirsute the CURRENT_TAGS property seems to be updated correctly when rules are removed. As expected, the TAGS property remains tagged for the snap even after rules were removed.

However, I have noticed a weird behavor when disconnecting the interface. The tag was not removed from CURRENT_TAGS until I ran `udevadm trigger && udevadm settle`. Looking at our udev backend code, we call `b.reloadRules(nil)` on remove, which if I'm reading the code right, does not trigger the input subsystem due to --subsystem-nomatch=input, with the execption of joystick rule which is done towards the end of the function.

**Comment by bboozzoo** [comment]
Checked on 20.04, upon disconnecting the interface the tag is not removed. However, the tag gets removed when I do manual `udevadm trigger && udevadm settle`, which suggests the backend reloadRules() does not trigger the input subsystem correctly when disconnecting an interface.

**Comment by anonymouse64** [comment]
Nice detective work on that @bboozzoo ! I think in the short term we still probably want to land this as-is but it sounds like we may need some additional snap-confine and/or snapd interface backend work to ensure that removing/untagging devices works the way we want it to on hirsute and forward.

**Pull request merged**